### PR TITLE
[Compression] lightning & legacy evaluator - step 1

### PR DESCRIPTION
**Pull request by J-shang**
### Description ###
The purpose of these series of prs is to support lightning and make training-related logic during compression modular. One who uses lightning could directly use nni compression without changing their model or training code.

- step 1: add evaluator & support lr scheduler
- step 2: let pruner & experiment using evaluator

#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [x] test case
  - [ ] doc

### CODE REVIEW

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+        raise NotImplementedError
+
+
+class LightningEvaluator(Evaluator):
+    """
+    Parameters
+    ----------
+    trainer
+        Pytorch-Lightning Trainer.
+        If the test metric is needed by nni, please make sure log metric with key ``NNI_METRIC``.
```
Comment (ultmaster) [inline nni/algorithms/compression/v2/pytorch/utils/evaluator.py:147]: Looks like a sentence appearing in a wrong place.

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py [hunk 2]
```diff
@@ -0,0 +1,581 @@
...
+    def remove(self):
+        if self.handle is None:
+            _logger.warning('%s for %s has not been registered yet.', self.__class__.__name__, self.target_name)
+            return
+        self.handle.remove()
+        self.handle = None
+        self.buffer = []
+
+
+class TensorHook(Hook):
```
Comment (ultmaster) [inline nni/algorithms/compression/v2/pytorch/utils/evaluator.py:66]: What is a tensor hook?

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py [hunk 3]
```diff
@@ -0,0 +1,581 @@
...
+                        loss.backward()
+                        optimizer.step()
+                        current_steps += 1
+                        if max_steps and current_steps == max_steps:
+                            return
+                    scheduler.step()
+
+    optimziers
+        The traced optimizer instance which the optimizer class is wrapped by nni.trace.
+        E.g. ``traced_optimizer = nni.trace(torch.nn.Adam)(model.parameters())``.
```
Comment (ultmaster) [inline nni/algorithms/compression/v2/pytorch/utils/evaluator.py:428]: This will all appear in one line in the docs. Suggest numbering them with 1,2,3,4...

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py [hunk 4]
```diff
@@ -0,0 +1,581 @@
...
+            self.trainer.fit_loop.max_epochs = max_epochs
+        self.trainer.fit(self.model, self.data_module)
+        self.trainer.fit_loop.max_steps, self.trainer.fit_loop.max_epochs = ori_max_steps, ori_max_epochs
+
+    def finetune(self):
+        self.train()
+
+    def evaluate(self) -> Tuple[float | None, List[Dict[str, float]]]:
+        """
+        NNI will use metric with key ``NNI_METRIC`` for evaluating model, please make sure you have this key in your ``Trainer.test()`` returned metric dicts.
```
Comment (ultmaster) [inline nni/algorithms/compression/v2/pytorch/utils/evaluator.py:357]: `NNI_METRIC` -> `default`.

If there is only one metric, this is not necessary?

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py (hunk 3 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/utils/evaluator.py:428]: how to use 1,2,3?

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py (hunk 2 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/utils/evaluator.py:66]: use to hook the tensor grad https://pytorch.org/docs/stable/generated/torch.Tensor.register_hook.html

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py (hunk 1 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/utils/evaluator.py]: move to class docstring

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py [hunk 5]
```diff
@@ -0,0 +1,581 @@
...
+
+        self.model: pl.LightningModule | None = None
+        self._hooks: List[Hook] = []
+        self._ori_model_attr = {}
+        self._param_names_map: Dict[str, str] | None = None
+
+        self._initialization_complete = False
+
+    def _init_optimizer_helpers(self, pure_model: pl.LightningModule):
+        assert self._initialization_complete is False, 'Evaluator initialization is already complete.'
```
[the remark this answers is not in the document]
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/utils/evaluator.py:180]: no, it's nni logic

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py [hunk 6]
```diff
@@ -0,0 +1,581 @@
...
+    def patch_optimizer_step(self, before_step_tasks: List[Callable], after_step_tasks: List[Callable]):
+        # Run tasks in `before_step_tasks` before `optimizer.step()` each time.
+        # Run tasks in `after_step_tasks` after `optimizer.step()` each time.
+        # NOTE: we only patch these tasks to the first optimizer right now.
+        raise NotImplementedError
+
+    def revert_optimizer_step(self):
+        raise NotImplementedError
+
+    def register_hooks(self, hooks: List[Hook]):
```
[the remark this answers is not in the document]
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/utils/evaluator.py:121]: add a few docstrings...but I think we need write a doc for how to use evaluator in another pr.

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py [hunk 7]
```diff
@@ -0,0 +1,581 @@
...
+        If the test metric is needed by nni, please make sure log metric with key ``NNI_METRIC``.
+    data_module
+        Pytorch-Lightning LightningDataModule.
+    dummy_input
+        The dummy_input is used to trace the graph. If dummy_input is not given, will use the data in data_module.train_dataloader().
+    """
+
+    def __init__(self, trainer: pl.Trainer, data_module: pl.LightningDataModule,
+                 dummy_input: Any | None = None):
+        assert isinstance(trainer, pl.Trainer) and is_traceable(trainer)
```
[the remark this answers is not in the document]
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/utils/evaluator.py]: good suggestion, added

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py (hunk 3 above)
Comment (ultmaster) [reply nni/algorithms/compression/v2/pytorch/utils/evaluator.py:444]: Like in rst.

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ultmaster) [reply nni/algorithms/compression/v2/pytorch/utils/evaluator.py:183]: I mean the 6 cases of optimizer/scheduler.

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py (hunk 2 above)
Comment (ultmaster) [reply nni/algorithms/compression/v2/pytorch/utils/evaluator.py:66]: Why do we have a base class called Hook? What will happen if users write a customized hook following the tutorial you provided above?

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ultmaster) [reply nni/algorithms/compression/v2/pytorch/utils/evaluator.py:121]: Docstrings are docstrings. Tutorials and tutorials. They are different things. You can have a doc to educate beginners how to use the evaluator, but for advanced users and developers (like me), they still need to know what's the purpose of every API, and why you design them.

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py (hunk 2 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/utils/evaluator.py:66]: This is an internal util class, use to save some hook information (name, buffer, handle...), I think users will never this

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/utils/evaluator.py:183]: yes, here https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#configure-optimizers

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py (hunk 2 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/utils/evaluator.py:66]: will add some examples in docstrings for these `XXHook`

File: test/algo/compression/v2/test_evaluator.py [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+    pl_model = SimpleLightningModel()
+    pl_trainer = nni.trace(pl.Trainer)(
+        max_epochs=1,
+        max_steps=10,
+        logger=TensorBoardLogger(Path(__file__).parent / 'lightning_logs', name="resnet"),
+    )
+    pl_trainer.num_sanity_val_steps = 0
+    pl_data = nni.trace(MNISTDataModule)(data_dir=Path(__file__).parent / 'data')
+    evaluator = LightningEvaluator(pl_trainer, pl_data)
+    evaluator._init_optimizer_helpers(pl_model)
```
Comment (QuanluZhang) [inline test/algo/compression/v2/test_evaluator.py:250]: `_init_optimizer_helpers` is public method or internal method?

Comment (QuanluZhang) [reply test/algo/compression/v2/test_evaluator.py:250]: ok, it should be internal, never mind

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py [hunk 9]
```diff
@@ -0,0 +1,649 @@
...
+
+    def _register(self, hook_func: Callable[[Module, Tensor, Tensor], Any]):
+        return self.target.register_backward_hook(hook_func)  # type: ignore
+
+
+class Evaluator:
+    # A flag to indicate whether the evaluator is initialized complete.
+    _initialization_complete: bool
+
+    def _init_optimizer_helpers(self, pure_model: Module | pl.LightningModule):
```
Comment (ultmaster) [inline nni/algorithms/compression/v2/pytorch/utils/evaluator.py:122]: I think this also needs to be documented.

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ultmaster) [reply nni/algorithms/compression/v2/pytorch/utils/evaluator.py:151]: Who will use this?

**Review by ultmaster — APPROVED** [review]

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py [hunk 10]
```diff
@@ -0,0 +1,713 @@
...
+        Bind the model suitable for this ``Evaluator`` to use the evaluator's abilities of model modification, model training, and model evaluation.
+
+        Parameter
+        ---------
+        model
+            The model bind to this ``Evaluator``, usually a wrapped model.
+        param_names_map
+            ``param_names_map`` maps the names of the parameters in the pure_model to the names of the parameters in the bound model.
+            The format of param_names_map is {pure_model_param_name: bound_model_param_name}.
+            It is for initializing the optimizers for the bound model.
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/utils/evaluator.py:153]: this function is for initializing the optimizer?

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py [hunk 11]
```diff
@@ -0,0 +1,713 @@
...
+        This function is used to record the status of the optimizers & lr_schedulers,
+        and ensure NNI can reinitialize the optimizers & lr_schedulers with a similar but modified model.
+
+        NOTE: this is a part of Evaluator initialization, please make sure this function has been called before using other evaluator functions.
+        """
+        raise NotImplementedError
+
+    def bind_model(self, model: Module | pl.LightningModule, param_names_map: Dict[str, str] | None = None):
+        """
+        NOTE: NNI users don't need to call this function. This function usually called by NNI Compressor.
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/utils/evaluator.py:142]: I would suggest to remove all the NOTE message from member functions' docstring, only keep one in the class docstring

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py [hunk 12]
```diff
@@ -0,0 +1,713 @@
...
+        NOTE: NNI users are not recommended to call this function, it's used by NNI compressor.
+
+        Revert the loss to the original one.
+        """
+        raise NotImplementedError
+
+    def patch_optimizer_step(self, before_step_tasks: List[Callable], after_step_tasks: List[Callable]):
+        """
+        NOTE: NNI users are not recommended to call this function, it's used by NNI compressor.
+        NOTE: we only patch tasks to the first optimizer right now.
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/utils/evaluator.py:192]: why we do not need to patch tasks for our re-initialized optimizers?

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py [hunk 13]
```diff
@@ -0,0 +1,713 @@
...
+        if hasattr(self, '_hooks'):
+            for hook in self._hooks:
+                hook.remove()
+            self._hooks.clear()
+
+    def train(self, max_steps: int | None = None, max_epochs: int | None = None):
+        """
+        NOTE: NNI users can call this function to train the bound model, but not necessary.
+
+        Train the bound model with default optimization loop defined by user and only change the training duration.
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/utils/evaluator.py:244]: do we change the optimizer during training?

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py [hunk 14]
```diff
@@ -0,0 +1,713 @@
...
+        Finetune the bound model with default optimization loop defined by user.
+        """
+        raise NotImplementedError
+
+    def evaluate(self) -> float | Tuple[float, Any]:
+        """
+        NOTE: NNI users can call this function to evaluate the bound model, but not necessary.
+
+        Evaluate the bound model, the returned metric should be a float number,
+        or a tuple of two elements with the first element being a float number.
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/utils/evaluator.py:261]: "a tuple of two elements", what is the other element?

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py [hunk 15]
```diff
@@ -0,0 +1,713 @@
...
+        NOTE: NNI users are not recommended to call this function, it's used by NNI compressor.
+
+        The returned value is a dummy input for the model, always used by ``torch.jit.trace``.
+        """
+        raise NotImplementedError
+
+
+class LightningEvaluator(Evaluator):
+    """
+    NOTE: If the the test metric is needed by nni, please make sure log metric with key ``default`` in LightningModule.test_step().
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/utils/evaluator.py:278]: would be better to add short description for this class, then the NOTE message follows the description. for example, the description could be "LightningEvaluator is the Evaluator based on PytorchLightning. It is very friendly to the users who are familiar to PytorchLightning or already have training/validation/testing code written in PytorchLightning."

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py (hunk 13 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/utils/evaluator.py:244]: we might change optimizer, and this line means for this function, it can only change the duration.
If pruner/quantizer want to change optimizer, it should call `patch_optimizer_step`

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py (hunk 14 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/utils/evaluator.py:261]: The other element is usually the original metric, any type is ok.
in lightning evaluator, the second element is the returned value of `trainer.test()`, I will add more docstring to explain this.

File: nni/algorithms/compression/v2/pytorch/utils/evaluator.py (hunk 15 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/utils/evaluator.py:278]: good suggestion, will add them.

**Review by QuanluZhang — APPROVED** [review]
